tests/functionaltests/Tests/NSURLSession.mm

```diff
@@ -598,18 +598,21 @@ - (void)URLSession:(NSURLSession*)session downloadTask:(NSURLSessionDownloadTask
     LOG_INFO("Establishing download task with url %@", url);
     NSURLSessionDownloadTask* downloadTask = [session downloadTaskWithURL:url
                                                         completionHandler:^(NSURL* location, NSURLResponse* response, NSError* error) {
+                                                            [condition lock];
                                                             downloadLocation = location;
                                                             downloadResponse = response;
                                                             downloadError = error;
                                                             [condition signal];
+                                                            [condition unlock];
                                                         }];
     [downloadTask resume];
 
     // Wait for download to complete.
-    if (![condition waitUntilDate:[NSDate dateWithTimeIntervalSinceNow:c_testDownloadTimeoutInSec]]) {
-        // Wait timed out.
-        ASSERT_FALSE_MSG(true, "FAILED: Waiting for download timed out!");
-    }
+    [condition lock];
+    ASSERT_TRUE_MSG( (downloadLocation || downloadResponse || downloadError) || 
```

Comment:
> ASSERT_TRUE_MSG( (downloadLocation || downloadResponse || downloadError) ||  [](start = 3, length = 77)

some of these don't handle spurious wakeups well but its at least a lot better than before. NSCondition needs locking and if you signal before a wait it just goes into the ether so needed another piece of information to check against.